### ISSUE TITLE
__has_include with angle-bracketed argument breaks preprocessing

### ISSUE THREAD

**Issue by dummyunit** [danmar/simplecpp#229]
Preprocessing (in C++17 mode) of code like this:
```cpp
#if __has_include(<any>)
#endif
```
results in an error:
```
/tmp/fail.cpp:1: syntax error: failed to evaluate #if condition
```
This happens regardless of whether the mentioned header file exists or not.

Debugging shows that `<any>` is split in 3 tokens: `<`, `any`, `>`, but the code for handling `__has_include` expects one token as argument: https://github.com/danmar/simplecpp/blob/ab680f96ea9aa0fc7ed241f6dc51954d7a8d6735/simplecpp.cpp  
As a result the check to matching closing parenthesis fails (as `any` is not `)`) and a fatal syntax error is reported: https://github.com/danmar/simplecpp/blob/ab680f96ea9aa0fc7ed241f6dc51954d7a8d6735/simplecpp.cpp

This also affects Cppcheck - it reports `preprocessorErrorDirective` on that line and stops.

**Comment by danmar** [comment]
ouch I thought this __has_include was fixed by now. Thanks for reporting!

**Comment by jamesharris-garmin** [comment]
have we been able to fix this? or should someone submit a pull request?

**Comment by danmar** [comment]
I didn't have time to look at this yet. A PR would be appreciated.

> Follow up:

not sure I'd have to look in the code and/or debug..

**Comment by aremmell** [comment]
No movement? Does someone externally need to PR this to get it done? It's pretty bad! This is a super common construct to use these days

**Comment by danmar** [comment]
> No movement? Does someone externally need to PR this to get it done? It's pretty bad! This is a super common construct to use these days

This is an open source project. The development team is just a group of volounteers that helps out for free.

I am hoping that Cppcheck Premium will grow and that we can someday pay some developer to work on Simplecpp and Cppcheck..

**Comment by aremmell** [comment]
Sorry, didn't mean to come off the wrong way. I was just surprised about this one. I would try to fix it myself if I thought I could without breaking 8 other things.

I imagine the learning curve is quite steep if you've never done something similar before. I can give it a go though if you want :)

**Comment by danmar** [comment]
> Sorry, didn't mean to come off the wrong way.

ok.

> I imagine the learning curve is quite steep if you've never done something similar before. I can give it a go though if you want :)

This code is not super trivial but imho this is probably a relatively nice task to start working with compared to some other tasks. Feel free to take a look I would appreciate a PR!

**Comment by aremmell** [comment]
Where would one start looking?

**Comment by danmar** [comment]
spontanously I wonder what happens here:

https://github.com/danmar/simplecpp/blob/b9bfbaa4a98f9c2f9d8c41547488b146f0584837/simplecpp.cpp#L3494

at a quick glance it looks as if <> is handled.

**Comment by danmar** [comment]
in this test case it looks like <> are handled:

https://github.com/danmar/simplecpp/blob/b9bfbaa4a98f9c2f9d8c41547488b146f0584837/test.cpp#L1312

**Comment by danmar** [comment]
can somebody create a test case that fails? I don't get a syntax error here:

```
$ cat 229.cpp 
#if __has_include(<any>)
#endif
$ ./simplecpp 229.cpp 

$
```

**Comment by yut23** [comment]
The issues I was seeing with this seem to have been fixed by #271 (first included in cppcheck 2.10).

**Comment by danmar** [comment]
ok thanks.. I will close this.. if anybody still see problems feel free to reopen this issue or open a new issue..
